Mathlib/Data/Nat/Nth.lean

```diff
@@ -252,6 +252,59 @@ theorem le_nth_of_lt_nth_succ {k a : ℕ} (h : a < nth p (k + 1)) (ha : p a) : a
   · rcases subset_range_nth ha with ⟨n, rfl⟩
     rwa [nth_lt_nth hf, Nat.lt_succ_iff, ← nth_le_nth hf] at h
 
+lemma nth_add_one {n : ℕ} (h0 : ¬p 0) (h : nth p n ≠ 0) :
+    nth (fun i ↦ p (i + 1)) n + 1 = nth p n := by
```

Comment:
It does seem surprising that there isn't an easier proof here.

I wonder if generalising to `f (nth (p ∘ f) k) = nth (fun i => p i ∧ i ∈ image f) k` for an arbitrary order embedding `f : ℕ → ℕ` would end up being easier.

Reply:
That doesn't seem like it would be any easier to me. Maybe there's some kind of induction principle for `nth` that could abstract proofs of such things, though I suspect such a principle would be awkward to use in practice.

Reply:
I've refactored to have such a version with `+ m` (which may well introduce fresh scope for golfing the proof again). This isn't the fully general version with an order embedding (where `h0 : ∀ k, p k → k ∈ Set.range f` would be more closely analogous to these lemmas than the suggested version with a different `p` on the RHS).

Reply:
I've now pushed a more general `nth_comp_of_strictMono` lemma for order embeddings (which also uses the more general but more awkward `∀ hfi : (setOf p).Finite, n < hfi.toFinset.card` in place of `nth p n ≠ 0`) - this is my version with the same `p` on both sides, not the one suggested with a different `p` on the RHS. That change did make the file shorter, but there is probably again scope for more golfing.